src/syntax.ml

```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
+          | ':', '-' -> Some `Left
+          | '-', ':' -> Some `Right
+          | '-', '-' -> Some `Default
+          | _ -> None
+        else None
+    | _ -> None
+  in
+  let valid_align_row = function
+    | { Loc.value = `Word w; location } ->
+        let w =
+          if String.get w 0 = '|' then String.sub w 1 (String.length w - 1)
+          else w
+        in
+        let w =
+          if String.get w (String.length w - 1) = '|' then
+            String.sub w 0 (String.length w - 1)
+          else w
+        in
+        let ws = String.split_on_char '|' w in
+        let ws = List.map (fun x -> { Loc.value = `Word x; location }) ws in
+        if List.for_all (fun x -> Option.is_some (valid_align x)) ws then
+          (* revert the list for consistency *)
+          Some (List.map (fun x -> [ x ]) ws |> List.rev)
+        else None
+    | _ -> None
+  in
+  let rec consume_elements acc_t acc_r acc_c ~new_line ~last_loc =
+    let next_token = peek input in
+    match next_token.value with
+    | `Right_brace ->
+        junk input;
+        (acc_t, acc_r, acc_c, next_token.location)
+    | `Space _ ->
+        junk input;
+        consume_elements acc_t acc_r acc_c ~new_line ~last_loc
+    | `Single_newline _ | `Blank_line _ ->
+        junk input;
+        let acc_r =
+          match acc_c with [] -> acc_r | _ -> List.rev acc_c :: acc_r
+        in
+        let acc_t =
+          match acc_r with [] -> acc_t | _ -> List.rev acc_r :: acc_t
+        in
+        consume_elements acc_t [] [] ~new_line:true ~last_loc
+    | `Bar ->
+        junk input;
+        if new_line then consume_elements acc_t [] [] ~new_line:false ~last_loc
+        else
+          consume_elements acc_t (List.rev acc_c :: acc_r) [] ~new_line:false
+            ~last_loc
+    | #token_that_always_begins_an_inline_element as token ->
+        let i = inline_element input next_token.location token in
+        if new_line then
+          match valid_align_row i with
+          | Some align_row ->
+              consume_elements acc_t align_row [] ~new_line:false
+                ~last_loc:next_token.location
+          | None ->
+              consume_elements acc_t acc_r (i :: acc_c) ~new_line:false
+                ~last_loc:next_token.location
+        else
+          consume_elements acc_t acc_r (i :: acc_c) ~new_line:false
+            ~last_loc:next_token.location
+    | other_token ->
+        Parse_error.not_allowed
+          ~what:(Token.describe other_token)
+          ~in_what:(Token.describe parent_markup)
+          next_token.location
+        |> add_warning input;
+        junk input;
+        (acc_t, acc_r, acc_c, last_loc)
+  in
+  let acc_t, acc_r, acc_c, brace_location =
+    consume_elements [] [] [] ~new_line:true ~last_loc:parent_markup_location
+  in
+  let acc_r = match acc_c with [] -> acc_r | _ -> List.rev acc_c :: acc_r in
+  let acc_t = match acc_r with [] -> acc_t | _ -> List.rev acc_r :: acc_t in
+  let rows = List.rev acc_t in
```

Comment:
`acc_t`, `acc_r`, `acc_c` that's too hard to follow and adds unnecessary constraints: separator row validated once then parsed again, "impossible" cases, duplicated code.

I suggest making two "consume" functions: One for the header and separator, the other for the remaining rows. That way, the parsing of the first two is not mentioned all over the place and can return the more interesting type directly.
I believe this can also help to support an optional header row, and remove the duplicated error checking.

Reply:
I've split the consume function in two. But if we would change the design to first expect the header, then the align row, then the data it would not allow to have tables containing a single line, or only data, etc. I find it convenient to have the align row have the same type as the data (represented as Word) and interpret it when we parsed everything (allow more flexible tables). The code has been heavily modified, what do you think?